lib/publisher/auditpublisher.go

```diff
@@ -0,0 +1,105 @@
+package publisher
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/cloudflare/cfssl/log"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/trace"
+)
+
+const (
+	//default sqs queue name
+	defaultAuditSQSQueueName = "remote-access-audit-event-notification-queue"
+)
+
+var (
+	eventsToPublish = getSupportedAuditTypes()
+)
+
+type AuditPublisher interface {
+	apievents.Emitter
+}
+
+type AuditPublisherConfig struct {
+	Enabled      bool
+	SQSQueueName string
+}
+
+type AuditPublisherService struct {
+	publisher AuditMessagePublisher
+}
+
+type AuditMessage struct {
+	MessageBody string
+}
+
+// AuditMessagePublisher responsible for publish the audit message to destination
+type AuditMessagePublisher interface {
+	Publish(audit AuditMessage) error
+}
+
+func (cfg *AuditPublisherConfig) CheckAndSetDefaults() error {
+	if cfg.SQSQueueName == "" {
+		cfg.SQSQueueName = defaultAuditSQSQueueName
+	}
+	return nil
+}
+
+func NewAuditPublisher(cfg AuditPublisherConfig) AuditPublisher {
+	if cfg.Enabled {
+		log.Info("Audit publishing is eanbled")
+		cfg.CheckAndSetDefaults()
+		return &AuditPublisherService{
+			publisher: NewAuditMessagePublisherService(cfg),
+		}
+	}
+	log.Info("Audit publishing is disabled")
+	return &noOpAuditPublisher{}
+}
+
+// EmitAuditEvent implements AuditPublisher
+func (s *AuditPublisherService) EmitAuditEvent(ctx context.Context, event apievents.AuditEvent) error {
+	//filter the required event to publish
+	if _, ok := eventsToPublish[event.GetType()]; !ok {
+		log.Infof("Skipping the audit event id:%v type: %v", event.GetID(), event.GetType())
+		return nil
+	}
+
+	log.Infof("Publishing the audit event [ id %v  cluster: %v and type: %v ]", event.GetID(), event.GetClusterName(), event.GetType())
+	messageBody, err := json.Marshal(event)
+	if err != nil {
+		log.Errorf("Failed to serialize the audit event to json err: %v", err)
+		return trace.Wrap(err)
+	}
+
+	err = s.publisher.Publish(AuditMessage{
+		MessageBody: string(messageBody),
+	})
```

Comment:
When publishing use tenant name as message group id header so that we can enable scalability in future.
We need tenant separation

Reply:
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/using-messagegroupid-property.html

Reply:
Its not very clear from the aws docs whether its applicable for standard SQS queue or not.

As per the Go SDK this parameter is not applied to standard sqs queue

	// This parameter applies only to FIFO (first-in-first-out) queues.
	//
	// The tag that specifies that a message belongs to a specific message group.
	// Messages that belong to the same message group are processed in a FIFO manner
	// (however, messages in different message groups might be processed out of
	// order). To interleave multiple ordered streams within a single queue, use
	// MessageGroupId values (for example, session data for multiple users). In
	// this scenario, multiple consumers can process the queue, but the session
	// data of each user is processed in a FIFO fashion.
	//
	//    * You must associate a non-empty MessageGroupId with a message. If you
	//    don't provide a MessageGroupId, the action fails.
	//
	//    * ReceiveMessage might return messages with multiple MessageGroupId values.
	//    For each MessageGroupId, the messages are sorted by time sent. The caller
	//    can't specify a MessageGroupId.
	//
	// The length of MessageGroupId is 128 characters. Valid values: alphanumeric
	// characters and punctuation (!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~).
	//
	// For best practices of using MessageGroupId, see Using the MessageGroupId
	// Property (https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/using-messagegroupid-property.html)
	// in the Amazon SQS Developer Guide.
	//
	// MessageGroupId is required for FIFO queues. You can't use it for Standard
	// queues.
	MessageGroupId *string `type:"string"`